pixeltable/catalog/catalog.py

```diff
@@ -723,17 +823,31 @@ def get_view_ids(self, tbl_id: UUID, for_update: bool = False) -> list[UUID]:
         return result
 
     def get_tbl_version(self, tbl_id: UUID, effective_version: Optional[int]) -> Optional[TableVersion]:
-        if (tbl_id, effective_version) not in self._tbl_versions:
-            self._tbl_versions[tbl_id, effective_version] = self._load_tbl_version(tbl_id, effective_version)
-        return self._tbl_versions[tbl_id, effective_version]
-
-    def add_tbl_version(self, tbl_version: TableVersion) -> None:
-        """Explicitly add a TableVersion"""
-        self._tbl_versions[tbl_version.id, tbl_version.effective_version] = tbl_version
-        # if this is a mutable view, also record it in the base
-        if tbl_version.is_view and tbl_version.effective_version is None:
-            base = tbl_version.base.get()
-            base.mutable_views.append(TableVersionHandle(tbl_version.id, tbl_version.effective_version))
+        # we need a transaction here, if we're not already in one; if this starts a new transaction,
+        # the returned TableVersion instance will not be validated
+        with self.begin_xact(tbl_id=tbl_id, for_write=False) as conn:
+            tv = self._tbl_versions.get((tbl_id, effective_version))
+            if tv is None:
+                tv = self._load_tbl_version(tbl_id, effective_version)
+            elif not tv.is_validated and effective_version is None:
+                # we validate live instances by comparing our cached version number to the stored current version
+                _logger.debug(f'validating metadata for table {tbl_id}:{tv.version}')
+                q = sql.select(schema.Table.md).where(schema.Table.id == tbl_id)
+                row = conn.execute(q).one()
+                current_version = row.md['current_version']
+
+                # the stored version can be behind TableVersion.version, because we don't roll back the in-memory
+                # metadata changes after a failed update operation
+                if current_version != tv.version:
+                    # the cached metadata is invalid
+                    _logger.debug(
+                        f'reloading metadata for table {tbl_id} '
+                        f'(cached version: {tv.version}, current version: {current_version})'
+                    )
+                    tv = self._load_tbl_version(tbl_id, effective_version)
+
+            tv.is_validated = True
```

Comment:
It's not clear to me why this is necessary. Shouldn't it be the case that if `tv` was invalid, we created a new one, which is necessarily valid? Maybe I'm missing something here.

Reply:
superfluous, removed.